graphql-kotlin-federation/src/test/kotlin/com/expediagroup/graphql/federation/FederatedSchemaValidatorTest.kt

```diff
@@ -0,0 +1,226 @@
+
+    /**
+     * type Foo {}
+     */
+    @Test
+    fun `validate GraphQLObjectType but not federated should do nothing`() {
+        val typeToValidate = GraphQLObjectType.newObject()
+            .name("Foo")
+            .build()
+
+        assertDoesNotThrow {
+            FederatedSchemaValidator().validateGraphQLType(typeToValidate)
+        }
+    }
+
+    /**
+     * interface Foo {
+     *   bar: String
+     * }
+     */
+    @Test
+    fun `validate GraphQLInterfaceType but not federated should do nothing`() {
+        val typeToValidate = GraphQLInterfaceType.newInterface()
+            .name("Foo")
+            .field(GraphQLFieldDefinition.newFieldDefinition().name("bar").type(GraphQLString))
+            .build()
+
+        assertDoesNotThrow {
+            FederatedSchemaValidator().validateGraphQLType(typeToValidate)
+        }
+    }
+
+    /**
+     * type Foo @extends {}
+     */
+    @Test
+    fun `validate federated GraphQLObjectType with no fields`() {
+        val typeToValidate = GraphQLObjectType.newObject()
+            .name("Foo")
+            .withDirective(extendsDirectiveType)
+            .build()
+
+        assertFailsWith<InvalidFederatedSchema> {
+            FederatedSchemaValidator().validateGraphQLType(typeToValidate)
+        }
+    }
+
+    /**
+     * type Foo @extends {
+     *   bar: String
+     * }
+     */
+    @Test
+    fun `validate federated GraphQLObjectType with no key directive`() {
+        val field = GraphQLFieldDefinition.newFieldDefinition()
+            .name("bar")
+            .type(GraphQLString)
+
+        val typeToValidate = GraphQLObjectType.newObject()
+            .name("Foo")
+            .field(field)
+            .withDirective(extendsDirectiveType)
+            .build()
+
+        assertFailsWith<InvalidFederatedSchema> {
+            FederatedSchemaValidator().validateGraphQLType(typeToValidate)
+        }
+    }
+
+    /**
+     * type Foo @key {
+     *   bar: String
+     * }
+     */
+    @Test
+    fun `validate federated GraphQLObjectType with key directive`() {
+        val field = GraphQLFieldDefinition.newFieldDefinition()
+            .name("bar")
+            .type(GraphQLString)
+
+        val typeToValidate = GraphQLObjectType.newObject()
+            .name("Foo")
+            .field(field)
+            .withDirective(keyDirectiveType)
+            .build()
+
+        val result = kotlin.runCatching {
+            FederatedSchemaValidator().validateGraphQLType(typeToValidate)
+        }
+
+        val expectedError = """
+            Invalid federated schema:
+             - @key directive on Foo is missing field information
+        """.trimIndent()
+
+        assertEquals(expectedError, result.exceptionOrNull()?.message)
+    }
+
+    /**
+     * type Foo @key {
+     *   bar: String @provides
+     * }
+     */
+    @Test
+    fun `validate federated GraphQLObjectType with key directive and field with provides directive`() {
```

Comment:
should this be testing valid `@key` directive and invalid `@provides` instead?

Reply:
I can do that, it requires just a little moving around to get a value into the directive builder